### PR TITLE
Deploy NFT contracts 

### PR DESCRIPTION
**Pull request by RafilxTenfen**
### Deploy NFT Bridge

Deployed NFT contracts and Federation upgrade

### Description

- Add `IFederation` interface
- Add deploy script for NFT contracts
  - NFT Side Token
  - NFT Bridge
  - NFT Bridge Proxy
  - Federation NFT upgrade
  - ERC721 token for test

### How to Test

- Deploy the contracts in hardhat

#### Case 1

1. Go to bridge directory
```shell
$~ cd bridge
```

2. Run the hardhat deploy
```shell
$~ npx hardhat deploy
```

__Expected Result__
- It should run all the files inside the `bridge/deploy` path

#### Bridge Directory `cd bridge`
- [x] Lint is clean `npm run lint`
- [x] Test is passing `npm run test`
- [x] Contracts are compiling `npm run compile`

#### Federator Directory `cd federator`
- [x] Lint is clean `npm run lint`
- [x] Test is passing `npm run test`
- [x] Typescript is compiling `npm run build`

### CODE REVIEW

File: bridge/contracts/Federation.sol [hunk 1]
```diff
@@ -280,6 +262,18 @@ contract Federation is Initializable, UpgradableOwnable {
         return processed[transactionId];
     }
 
+    /**
+      @notice Gets the hash of transaction from the following parameters encoded and keccaked
+      @dev Its encode and keccak256 the parameters received in the same order
+      @param originalTokenAddress The address of the token in the origin (main) chain
```
[the remark this answers is not in the document]
Comment (pmprete) [reply bridge/contracts/Federation.sol:268]: It's the original token address on the main chain. Before emiting the event we check if its a side token and then it uses the  original token address on the main chain.

**Review by pmprete — APPROVED** [review]
LGTM

File: federator/src/contracts/FederationFactory.js [hunk 2]
```diff
@@ -26,7 +26,7 @@ module.exports = class FederationFactory extends ContractFactory {
         let federationContract = this.getContractByAbi(abiFederationNew, address, web3);
         const version = await this.getVersion(federationContract);
 
-        if (version === 'v2') {
+        if (version === 'v2' || version === 'v3') {
```
Comment (pmprete) [inline federator/src/contracts/FederationFactory.js:29]: We want to move to version 3 and leave version 2 as the old one like this:
```if (version === 'v3') {
          return new FederationInterfaceV3(this.config, federationContract);
        } else if (version === 'v2') {
         return new FederationInterfaceV2(this.config, federationContract);```
       
     We should tackle this when we restructure the deploy of the old Contracts

Comment (RafilxTenfen) [reply federator/src/contracts/FederationFactory.js:29]: > ```
>      We should tackle this when we restructure the deploy of the old Contracts
> ```

agree

**Pull request merged**